src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -0,0 +1,235 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+#nullable disable
+
+using EnvDTE;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Exceptions;
+using Microsoft.VisualStudio.ProjectSystem.VS.References.Roslyn;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.References
+{
+    [Export(typeof(IProjectSystemReferenceCleanupService))]
+    internal partial class ReferenceCleanupService : IProjectSystemReferenceCleanupService
+    {
+        private readonly AbstractReferenceHandler _projectAbstractReferenceHandler = new ProjectAbstractReferenceHandler();
+        private readonly AbstractReferenceHandler _packageAbstractReferenceHandler = new PackageAbstractReferenceHandler();
+        private readonly AbstractReferenceHandler _assemblyAbstractReferenceHandler = new AssemblyAbstractReferenceHandler();
```

Comment:
I don't think you need these fields. In the methods below, can't you foreach over `_mapReferenceTypeToHandler.Values.Where(handler => handler != null)`

Reply:
done